### PR TITLE
[Feature] Add Cylinder3D head

### PR DESCRIPTION
**Pull request by xizaoqu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add Cylinder3D head

## Modification

Add Cylinder3D head

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tests/test_models/test_decode_heads/test_cylinder3d_head.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+from mmcv.ops import SparseConvTensor
+
+from mmdet3d.models.decode_heads import Cylinder3DHead
+from mmdet3d.structures import Det3DDataSample, PointData
+
+
+class TestCylinder3DHead(TestCase):
+
+    def test_cylinder3d_head_loss(self):
+        """Tests DGCNN head loss."""
```
Comment (Tai-Wang) [inline tests/test_models/test_decode_heads/test_cylinder3d_head.py:14]: Cylinder3D

File: tests/test_models/test_decode_heads/test_cylinder3d_head.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
```
Comment (Tai-Wang) [inline tests/test_models/test_decode_heads/test_cylinder3d_head.py:15]: if not torch.cuda.is_available():
            pytest.skip('test requires GPU and torch+cuda')

File: tests/test_models/test_decode_heads/test_cylinder3d_head.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from unittest import TestCase
+
+import torch
+from mmcv.ops import SparseConvTensor
+
+from mmdet3d.models.decode_heads import Cylinder3DHead
+from mmdet3d.structures import Det3DDataSample, PointData
+
+
+class TestCylinder3DHead(TestCase):
```
Comment (Tai-Wang) [inline tests/test_models/test_decode_heads/test_cylinder3d_head.py:11]: Also need a UT for lovasz loss (only for the added loss)

File: mmdet3d/models/losses/lovasz_loss.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+    @property
+    def loss_name(self):
+        """Loss Name.
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
+        name.
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/lovasz_loss.py:319]: Need a blank line above "Returns"

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                 conv_cfg: ConfigType = dict(type='Conv1d'),
+                 norm_cfg: ConfigType = dict(type='BN1d'),
+                 act_cfg: ConfigType = dict(type='ReLU'),
+                 loss_ce: ConfigType = dict(
+                     type='mmdet.CrossEntropyLoss',
+                     use_sigmoid=False,
+                     class_weight=None,
+                     loss_weight=1.0),
+                 loss_lovasz: ConfigType = dict(
+                     type='mmseg.LovaszLoss', loss_weight=1.0),
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:58]: Since we have removed mmsegmentation dependency, it seems we need to implement LovaszLoss in mmdet3d?

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            Defaults to dict(type='ReLU').
+        loss_ce (dict or :obj:`ConfigDict`): Config of CrossEntropy loss.
+            Defaults to dict(
+                     type='mmdet.CrossEntropyLoss',
+                     use_sigmoid=False,
+                     class_weight=None,
+                     loss_weight=1.0).
+        loss_lovasz (dict or :obj:`ConfigDict`): Config of Lovasz loss.
+            Defaults to dict(type='mmseg.LovaszLoss', loss_weight=1.0).
+        conv_seg_kernel_size (int): The kernel size used in conv_seg.
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:37]: Defaults to 3?

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            Defaults to 1.
+        ignore_index (int): The label index to be ignored. When using masked
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:39]: check the defaults number

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                     batch_data_samples: SampleList) -> dict:
+        """Compute semantic segmentation loss.
+
+        Args:
+            seg_logit (spconv.SparseConvTensor): Predicted per-voxel
+                segmentation logits of shape [num_voxels, num_classes]
+                stored in SparseConvTensor.
+            batch_data_samples (List[:obj:`Det3DDataSample`]): The seg
+                data samples. It usually includes information such
+                as `metainfo` and `gt_pts_seg`.
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:102]: Returns: ...

File: mmdet3d/models/decode_heads/cylinder3d_head.py (hunk 5 above)
Comment (ZCMax) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:58]: change 'mmseg.LovaszLoss' to 'LovaszLoss'

File: mmdet3d/models/losses/lovasz_loss.py [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmengine.utils import is_list_of
+
+from mmdet3d.registry import MODELS
+from .lovasz_loss_utils import get_class_weight, weight_reduce_loss
+
+
+def lovasz_grad(gt_sorted):
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:18]: add type hint for this part of code

File: tests/test_models/test_decode_heads/test_cylinder3d_head.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            loss_ce=dict(
+                type='mmdet.CrossEntropyLoss',
+                use_sigmoid=False,
+                class_weight=None,
+                loss_weight=1.0),
+            loss_lovasz=dict(
+                type='LovaszLoss', loss_weight=1.0, reduction='none'),
+        ).cuda()
+
+        # DGCNN head expects dict format features
```
Comment (ZCMax) [inline tests/test_models/test_decode_heads/test_cylinder3d_head.py:30]: Cylinder3D head

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                 loss_ce: ConfigType = dict(
+                     type='mmdet.CrossEntropyLoss',
+                     use_sigmoid=False,
+                     class_weight=None,
+                     loss_weight=1.0),
+                 loss_lovasz: ConfigType = dict(
+                     type='LovaszLoss', loss_weight=1.0),
+                 conv_seg_kernel_size: int = 3,
+                 ignore_index: int = 0,
+                 init_cfg=None) -> None:
```
Comment (ZCMax) [inline mmdet3d/models/decode_heads/cylinder3d_head.py:61]: add type hint here

File: mmdet3d/models/losses/lovasz_loss.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+    jaccard = 1. - intersection / union
+    if p > 1:  # cover 1-pixel case
+        jaccard[1:p] = jaccard[1:p] - jaccard[0:-1]
+    return jaccard
+
+
+def flatten_binary_logits(
+        logits: torch.Tensor,
+        labels: torch.Tensor,
+        ignore_index: int = None) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:44]: Optional[int] = None

File: mmdet3d/models/losses/lovasz_loss.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    """Flattens predictions and labels in the batch (binary case). Remove
+    tensors whose labels equal to 'ignore_index'.
+
+    Args:
+        probs (torch.Tensor): Predictions to be modified.
+        labels (torch.Tensor): Labels to be modified.
+        ignore_index (int | None): The label index to be ignored.
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:51]: (int, optional)

File: mmdet3d/models/losses/lovasz_loss.py [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+    perm = perm.data
+    gt_sorted = labels[perm]
+    grad = lovasz_grad(gt_sorted)
+    loss = torch.dot(F.relu(errors_sorted), grad)
+    return loss
+
+
+def lovasz_hinge(logits: torch.Tensor,
+                 labels: torch.Tensor,
+                 classes: str or List[int] = 'present',
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:125]: Union[str, List[int]]='present'

File: mmdet3d/models/losses/lovasz_loss.py [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+                 reduction: str = 'mean',
+                 avg_factor: int = None,
+                 ignore_index: int = 255) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): [B, H, W], logits at each pixel
+            (between -infty and +infty).
+        labels (torch.Tensor): [B, H, W], binary ground truth masks (0 or 1).
+        classes (str | list[int], optional): Placeholder, to be consistent with
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:137]: (Union[str, list[int]])

File: mmdet3d/models/losses/lovasz_loss.py [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+    grad = lovasz_grad(gt_sorted)
+    loss = torch.dot(F.relu(errors_sorted), grad)
+    return loss
+
+
+def lovasz_hinge(logits: torch.Tensor,
+                 labels: torch.Tensor,
+                 classes: str or List[int] = 'present',
+                 per_image: bool = False,
+                 class_weight: List[float] = None,
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:127]: Optional[List[float]]=None

File: mmdet3d/models/losses/lovasz_loss.py [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+                 ignore_index: int = 255) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): [B, H, W], logits at each pixel
+            (between -infty and +infty).
+        labels (torch.Tensor): [B, H, W], binary ground truth masks (0 or 1).
+        classes (str | list[int], optional): Placeholder, to be consistent with
+            other loss. Default: None.
+        per_image (bool, optional): If per_image is True, compute the loss per
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:139]: remove optional,  optional is only used for the parameter whose default value is None

File: mmdet3d/models/losses/lovasz_loss.py [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+            image instead of per batch. Default: False.
+        class_weight (list[float], optional): The weight for each class.
+            Default: None.
+        reduction (str, optional): The method used to reduce the loss. Options
+            are "none", "mean" and "sum". This parameter only works when
+            per_image is True. Default: 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. This parameter only works when per_image is True.
+            Default: None.
+        ignore_index (int | None): The label index to be ignored. Default: 255.
```
Comment (ZCMax) [inline mmdet3d/models/losses/lovasz_loss.py:242]: Unify the default format to `Defaults to xxx`

File: mmdet3d/models/losses/lovasz_loss.py [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+    vlabels = labels[valid]
+    return vprobs, vlabels
+
+
+def lovasz_hinge_flat(logits: torch.Tensor,
+                      labels: torch.Tensor) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): [P], logits at each prediction
```
[the remark this answers is not in the document]
Comment (xizaoqu) [reply mmdet3d/models/losses/lovasz_loss.py:103]: The shape of the tensor. I just copied it from mmseg.

File: mmdet3d/models/losses/lovasz_loss.py [hunk 20]
```diff
@@ -0,0 +1,398 @@
...
+    """
+
+    def __init__(self,
+                 loss_type: str = 'multi_class',
+                 classes: Union[str, List[int]] = 'present',
+                 per_sample: bool = False,
+                 reduction: str = 'mean',
+                 class_weight: Optional[Union[List[float], str]] = None,
+                 loss_weight: float = 1.0,
+                 loss_name: str = 'loss_lovasz'):
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/lovasz_loss.py:335]: In mmdet3d, do we need loss_name arguments?

File: mmdet3d/models/losses/lovasz_loss.py [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): Logits at each pixel
+            (between -infty and +infty) with shape [B, H, W].
+        labels (torch.Tensor): Binary ground truth masks (0 or 1)
+            with shape [B, H, W].
+        classes (Union[str, list[int]], optional): Placeholder, to be
+            consistent with other loss. Defaults to None.
+        per_sample (bool): If per_sample is True, compute the loss per
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/lovasz_loss.py:169]: per_sample -> samplewise.
If `samplewise=True`, compute and keep the loss for each sample, otherwise it will compute loss for the total batch

File: mmdet3d/models/losses/lovasz_loss.py [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+            are "none", "mean" and "sum". This parameter only works when
+            per_sample is True. Defaults to 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. This parameter only works when per_sample is True.
+            Defaults to None.
+        ignore_index (Union[int, None]): The label index to be ignored.
+            Defaults to 255.
+
+    Returns:
+        torch.Tensor: The calculated loss.
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/lovasz_loss.py:183]: Will the shape of the loss change according to samplewise=True or False?

File: mmdet3d/models/losses/lovasz_loss.py [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+        if class_weight is not None:
+            loss *= class_weight[c]
+        losses.append(loss)
+    return torch.stack(losses).mean()
+
+
+def lovasz_softmax(probs: torch.Tensor,
+                   labels: torch.Tensor,
+                   classes: Union[str, List[int]] = 'present',
+                   per_sample: bool = False,
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/lovasz_loss.py:250]: remember also revising the argument name here

File: mmdet3d/models/losses/lovasz_loss.py [hunk 24]
```diff
@@ -0,0 +1,398 @@
...
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_lovasz'.
+    """
+
+    def __init__(self,
+                 loss_type: str = 'multi_class',
+                 classes: Union[str, List[int]] = 'present',
+                 per_sample: bool = False,
+                 reduction: str = 'mean',
+                 class_weight: Optional[Union[List[float], str]] = None,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/losses/lovasz_loss.py:333]: In mmdet3d, we may do not support class_weight in str type.

File: mmdet3d/models/losses/lovasz_loss.py (hunk 21 above)
Comment (xizaoqu) [reply mmdet3d/models/losses/lovasz_loss.py:169]: It doesn't keep the loss for each sample, but stacks the losses and reduces them in `weight_reduce_loss`.

File: mmdet3d/models/losses/lovasz_loss.py (hunk 22 above)
Comment (xizaoqu) [reply mmdet3d/models/losses/lovasz_loss.py:183]: No, the loss will be reduced in the end.

File: mmdet3d/models/losses/lovasz_loss.py (hunk 20 above)
Comment (xizaoqu) [reply mmdet3d/models/losses/lovasz_loss.py:335]: I remove it.

File: mmdet3d/models/decode_heads/cylinder3d_head.py [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+            kernel_size=kernel_size,
+            stride=1,
+            padding=1,
+            bias=True)
+
+    def forward(self, sparse_voxels: SparseConvTensor) -> SparseConvTensor:
+        """Forward function."""
+        sparse_logits = self.cls_seg(sparse_voxels)
+        return sparse_logits
+
```
[the remark this answers is not in the document]
Comment (xizaoqu) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:94]: I implement it in `segmentor`, which one do you think it better?

[the remark this answers is not in the document]
Comment (Xiangxu-0103) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:94]: I suggest to map the logits (not predicted label) to point-level in `decode_head`.

[the remark this answers is not in the document]
Comment (xizaoqu) [reply mmdet3d/models/decode_heads/cylinder3d_head.py:93]: Done.

**Review by Xiangxu-0103 — APPROVED** [review]
